### DISCUSSION TITLE
Best Solution for Medium Quality Audio Streaming to ESP32

### DISCUSSION THREAD

**Discussion by jimruxton** [pschatzmann/arduino-audio-tools#1019] [Q&A] [answered]
I'm putting together a system for streaming audio from a raspberry pi with a connected audio interface to a number of ESP32s. On the pi side I'm using gstreamer to send the audio. On the ESP32 I'm using audio tools to capture the stream. Such an awesome tool set. I've been sending a broadcast UDP stream over wifi using a wifi router. It works well when I send a raw stream at 16 Kbs but I don't get a great distance before getting breaks in the audio,  maybe about 15 feet. I don't need a very high sample rate as I am mostly concerned with sending lower frequency audio. I figured if I lowered the bit rate I could get greater distance.  My latest attempt is to use AAC compression at 22050 bps. This worked better and I probably get about 30 feet but would love to get even better distance and reception. So far I am just using ESP32s with an internal antenna.I can't seem to find the mini ESP modules with external antenna.  Wondering if anyone has thoughts for improvement. Would another codec such as Opus give me a lower bit rate hence greater distance? I was wondering if I should try ESPnow, though I like the idea of using a router connected to the PI for sending audio. If anyone has thoughts for improvement I would love to hear your suggestions.

**Comment by pschatzmann** [comment, marked as answer]
A couple of thoughts:
- try to increase the transmission power on your wifi router
- there are quite a few esp32 with external antennas, so this is definitely worth a trial. e.g. [from seeedoino](https://wiki.seeedstudio.com/XIAO_ESP32C3_Getting_Started/)
- Opus might be interesting because it provides some configurable forward error correction and you can try to use udp to do the frameing to get away with the requirement for a container
- The root cause of your issue however is UDP (packet loss). So I would suggest to consider to look into some tcp  based streaming solution e.g. using http (e.g. icecast)

**Reply by jimruxton** [reply]
Thanks very much for these suggestions. Today I experimented with icecast and trying both darkice and liquidsoap to send a stream from jack. liquidsoap won giving me a clean stream. Though I haven't tried it yet I'm worried that when having multiple ESP32s receiving the stream they probably won't be in sync and there is quite a latency. Anyway at least I get a clean signal. Perhaps using an external antenna or opus will work and keep all the units in sync without using icecast.

**Comment by pschatzmann** [comment]
I thought it would make sense to have a RTSP client available as well. Quite some time ago I was trying to convert the [live555](https://github.com/pschatzmann/arduino-live555) for Arduino but could never make the streamers work properly.

This week I implemented a [RTSP client](https://github.com/pschatzmann/arduino-live555/blob/master/examples/Mp3Client/Mp3Client.ino) and this seems to work quite nicely

**Reply by jimruxton** [reply]
Thanks for this.I'll have to give it a try, I have had very good luck with the opus codec.  I tried adding the forward error correction but didn't get noticeable improvement. Maybe I was doing it wrong. Using Opus appears to work even better than the icecast solution. I haven't tried it yet on multiple devices  but since I'm sending a broadcast stream I assume they should all be in sync. I wish I had a bit less latency. I'm using the following Gstreamer pipeline to send the opus encoded stream. 

`gst-launch-1.0 -v jackaudiosrc connect=0 ! audioconvert ! audioresample ! opusenc bitrate=32000 complexity=5 max-payload-size=1000  !  udpsink host=192.168.1.255 port=8000
`
Bringing max-payload-size down didn't seem to affect latency as I thought it would. I can adjust the frame-size in ms. The default is 20ms. Putting it lower messes up the stream and causes glitches. Anyway it's working pretty well. Thanks again for such a great library.

**Reply by dominickairmyne** [reply]
Hi I am working on a a similar project any chance I could see your code? I would really appreciate it!
